### PR TITLE
Reimpl cformat to suit for bytes

### PR DESCRIPTION
**Pull request by qingshi163**

### CODE REVIEW

**Comment by qingshi163** [comment]
Fixes #2121

File: vm/src/cformat.rs [hunk 1]
```diff
@@ -412,6 +561,159 @@ impl CFormatPart {
     }
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct CFormatBytes {
+    parts: Vec<(usize, CFormatPart)>,
```
Comment (coolreader18) [inline vm/src/cformat.rs:566]: CFormatPart should probably be generic to allow `String` or `Vec<u8>`

Comment (qingshi163) [reply vm/src/cformat.rs:566]: I'm not sure that is necessary, I am using String as Vec<u8> here for bytes. As String stores char as utf-8 so I don't think any cost here.

File: vm/src/cformat.rs [hunk 2]
```diff
@@ -122,10 +120,36 @@ struct CFormatSpec {
...
+        consume_length(iter);
+        let (format_type, format_char) = parse_format_type(iter)?;
+        let precision = if precision.is_some() {
+            precision
+        } else {
+            match format_type {
+                CFormatType::Float(_) => Some(CFormatQuantity::Amount(6)),
+                _ => None,
+            }
+        };
```
Comment (youknowone) [inline vm/src/cformat.rs:141]:
```rust
let precision = precision.or_else(|| match format_type {
   CFormatType::Float(_) => Some(CFormatQuantity::Amount(6)),
   _ => None,
});
```

File: vm/src/cformat.rs (hunk 1 above)
Comment (youknowone) [reply vm/src/cformat.rs:566]: @coolreader18 do you have an example which distinguish bytes and str for the case?

Comment (coolreader18) [reply vm/src/cformat.rs:566]: Well bytes can have non-ascii and still be formatted, while a String wouldn't conserve that:

```py
>>> b'\xaa %b' % b'123'
b'\xaa 123'
```

Specifically, the `u8 as char` casting in this PR would map the non-ascii byte `\xaa` to the utf8 character sequence `'ª' == b'\xc2\xaa'`

**Review by coolreader18 — APPROVED** [review]
Looks great!!

**Pull request merged**